pageserver/src/tenant/size.rs

```diff
@@ -44,6 +44,129 @@ struct TimelineInputs {
+            // there is a chance that it is at the same Lsn
+            if !known_timelineids.contains(&parent_id) {
+                info!("Found possibly orphan branch {:?}", curr_upd);
+
+                let mut curr_lsn_branchfroms: HashMap<TimelineId, Vec<(TimelineId, usize)>> =
+                    HashMap::new();
+
+                // inspect all branchpoints at the same lsn
+                while j < updates.len() && updates[j].lsn == curr_upd.lsn {
+                    let lookahead_upd = &updates[j];
+                    j += 1;
+
+                    if let Command::BranchFrom(lookahead_parent_id) = lookahead_upd.command {
+                        match lookahead_parent_id {
+                            Some(lookahead_parent_id)
+                                if !known_timelineids.contains(&lookahead_parent_id) =>
+                            {
+                                // we have not processed ancestor yet
+                                // store it for later
+                                curr_lsn_branchfroms
+                                    .entry(lookahead_parent_id)
+                                    .and_modify(|e| {
+                                        e.push((lookahead_upd.timeline_id, j));
+                                    })
+                                    .or_insert_with(|| vec![(lookahead_upd.timeline_id, j)]);
+                            }
+                            _ => {
+                                // we have already processed ancestor
+                                // process this BranchFrom Update normally
+                                known_timelineids.insert(lookahead_upd.timeline_id);
+                                sorted_updates.push(*lookahead_upd);
+                            }
+                        }
+                    }
+                }
+
+                // process BranchFroms in the tree order
+                // check that we don't have a cycle if somet entry is orphan
+                // (this should not happen, but better to be safe)
+                let mut processed_some_entry = true;
+                while processed_some_entry {
+                    processed_some_entry = false;
+
+                    curr_lsn_branchfroms.retain(|parent_id, branchfroms| {
+                        if known_timelineids.contains(parent_id) {
+                            for (timeline_id, j) in branchfroms {
+                                known_timelineids.insert(*timeline_id);
+                                sorted_updates.push(updates[*j - 1]);
+                            }
+                            processed_some_entry = true;
+                            false
+                        } else {
+                            true
+                        }
+                    });
+                }
+
+                if !curr_lsn_branchfroms.is_empty() {
+                    anyhow::bail!("orphan branch detected in BranchFroms");
```

Comment:
re: orphans, I wonder if those are possible if the parent/ancestral timeline would be broken, then what would happen for the children? New cases probably have come up recently with the on-demand download, but I would suspect that the children are made broken if any parent is broken.

Hmm, I cannot see any new cases in the `test_regress/fixtures/test_broken_timeline.py`.

Reply:
Well, this should never happen during normal work. This code is just to exit early in case if anything is broken.
similar to  this `tree_sort_timeline` [error](https://github.com/neondatabase/neon/blob/main/pageserver/src/tenant.rs#L1593)

Reply:
Tried to look around, my best guess is that *if* we can get these orphans[^orphans] they are something a transient failure case which I guess will need to reload the tenant manually or restart pageserver. Then it's probably ok to fail the size calculation.

[^orphans]: assuming my understanding is correct about what they would be, above comment.

Reply:
Reopening this discussion.

I think we should log the orphan, just in case as it will be interesting context. The `tenant_id` is already part of the span on `gather_size_inputs` so that wont be necessary.

```suggestion
                    anyhow::bail!("orphan branch(es) detected in BranchFroms: {curr_lsn_branchfroms:?}");
```

This should give us at least some context to work with. Most likely the message wont be huge.